crates/claims/crates/vc/src/syntax/non_empty_vec.rs

```diff
@@ -0,0 +1,90 @@
+use serde::{Deserialize, Serialize};
+use std::ops::Deref;
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+#[serde(try_from = "Vec<T>", into = "Vec<T>")]
+pub struct NonEmptyVec<T: Clone>(Vec<T>);
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("expected a non-empty array")]
+    Empty,
+}
+
+impl<T: Clone> NonEmptyVec<T> {
+    pub fn new(t: T) -> Self {
+        Self(vec![t])
+    }
+
+    pub fn maybe_new(v: Vec<T>) -> Option<Self> {
+        Self::try_from(v).ok()
+    }
+
+    pub fn push(&mut self, t: T) {
+        self.0.push(t)
+    }
+
+    pub fn into_inner(self) -> Vec<T> {
+        self.0
+    }
+
+    pub fn into<T2>(self) -> NonEmptyVec<T2>
```

Comment:
That will clash with the `Into::into` function, where the intent may be to convert a `NonEmptyVec` into a `Vec`. So I would suggest finding another name for this, like `map` (although this is usually for functions accepting a mapping function as parameter), or `cast`.

Reply:
Yes. Considering `Vec` doesn't itself implement such functions, I've decided to remove them in favour of implementing `IntoIter` so a consumer can easily do the same themselves

Reply:
Can you also add `impl<'a, T> IntoIterator for &'a mut NonEmptyVec<T> {}`?